hail/src/main/scala/is/hail/types/physical/PNDArray.scala

```diff
@@ -34,11 +34,55 @@ abstract class PNDArray extends PType {
 
   def numElements(shape: IndexedSeq[Value[Long]], mb: EmitMethodBuilder[_]): Code[Long]
 
-  def makeShapeBuilder(shapeArray: IndexedSeq[Value[Long]]): StagedRegionValueBuilder => Code[Unit]
+  def makeShapeBuilder(shapeArray: IndexedSeq[Value[Long]]): StagedRegionValueBuilder => Code[Unit] = { srvb =>
+    coerce[Unit](Code(
+      srvb.start(),
+      Code(shapeArray.map(shapeElement => Code(
+        srvb.addLong(shapeElement),
+        srvb.advance()
+      )))
+    ))
+  }
 
-  def makeRowMajorStridesBuilder(sourceShapeArray: IndexedSeq[Value[Long]], mb: EmitMethodBuilder[_]): StagedRegionValueBuilder => Code[Unit]
+  def makeColumnMajorStridesBuilder(sourceShapeArray: IndexedSeq[Value[Long]], mb: EmitMethodBuilder[_]): StagedRegionValueBuilder => Code[Unit] = { srvb =>
+    val runningProduct = mb.newLocal[Long]()
+    val tempShapeStorage = mb.newLocal[Long]()
+    Code(
+      srvb.start(),
+      runningProduct := elementType.byteSize,
+      Code.foreach(0 until nDims){ index =>
+        Code(
+          srvb.addLong(runningProduct),
+          srvb.advance(),
+          tempShapeStorage := sourceShapeArray(index),
+          runningProduct := runningProduct * (tempShapeStorage > 0L).mux(tempShapeStorage, 1L)
+        )
+      }
+    )
+  }
 
-  def makeColumnMajorStridesBuilder(sourceShapeArray: IndexedSeq[Value[Long]], mb: EmitMethodBuilder[_]): StagedRegionValueBuilder => Code[Unit]
+  def makeRowMajorStridesBuilder(sourceShapeArray: IndexedSeq[Value[Long]], mb: EmitMethodBuilder[_]): StagedRegionValueBuilder => Code[Unit] = { srvb =>
+    val runningProduct = mb.newLocal[Long]()
+    val tempShapeStorage = mb.newLocal[Long]()
+    val computedStrides = (0 until nDims).map(_ => mb.genFieldThisRef[Long]())
+    Code(
+      srvb.start(),
+      runningProduct := elementType.byteSize,
+      Code.foreach((nDims - 1) to 0 by -1){ index =>
+        Code(
+          computedStrides(index) := runningProduct,
+          tempShapeStorage := sourceShapeArray(index),
+          runningProduct := runningProduct * (tempShapeStorage > 0L).mux(tempShapeStorage, 1L)
+        )
+      },
+      Code.foreach(0 until nDims)(index =>
+        Code(
+          srvb.addLong(computedStrides(index)),
+          srvb.advance()
+        )
+      )
+    )
+  }
 
```

Comment:
These utility methods were lifted off of `PCanonicalNDArray` because they're really just helper methods to make SRVBs in a way that seems useful beyond just the canonical ndarray.